File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0127_Word_Ladder.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class Question_0127_Word_Ladder {
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        Set<String> begin = new HashSet<>(Arrays.asList(beginWord));
        Set<String> end = new HashSet<>(Arrays.asList(endWord));
        Set<String> bank = new HashSet<>(wordList);
        if (!bank.contains(endWord)) {
            return 0;
        }
        return help(begin, end, bank, 1);
    }
    
    private int help(Set<String> begin, Set<String> end, Set<String> bank, int level) {
        if (begin.size() == 0) {
            return 0;
        }
        level++;
        Set<String> n = new HashSet<>();
        for (String s : begin) {
            char[] arr = s.toCharArray();
            for (int i = 0; i < arr.length; i++) {
                char t = arr[i];
                for (char c = 'a'; c <= 'z'; c++) {
                    if (c != t) {
                        arr[i] = c;
                        String cur = new String(arr);
                        if (end.contains(cur)) {
                            return level;
                        }
                        if (bank.contains(cur)) {
                            n.add(cur);
                            bank.remove(cur);
                        }
                    }
                }
                arr[i] = t;
            }
        }
        return n.size() > end.size() ? help(end, n, bank, level) : help(n, end, bank, level);
    }
}

File: algorithms/algorithms-cplusplus/leetcode/Question_0127_Word_Ladder.cc
#include <vector>
#include <string>
#include <unordered_set>

using namespace std;

class Solution {
 public:
  int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
    unordered_set<string> begin({beginWord}), end({endWord}),
        bank(wordList.begin(), wordList.end());
    if (bank.find(endWord) == bank.end()) {
      return 0;
    }
    return help(begin, end, bank, 1);
  }

  int help(unordered_set<string>& begin, unordered_set<string>& end,
      unordered_set<string>& bank, int level) {
    if (begin.empty()) {
      return 0;
    }
    level += 1;
    unordered_set<string> newBegin;
    for (string s : begin) {
      for (size_t i = 0; i < s.size(); i++) {
        char tem = s[i];
        for (char ch = 'a'; ch <= 'z'; ch++) {
          if (ch != tem) {
            s[i] = ch;
            if (end.find(s) != end.end()) {
              return level;
            }
            if (bank.find(s) != bank.end()) {
              newBegin.insert(s);
              bank.erase(s);
            }
          }
        }
        s[i] = tem;
      }
    }

    if (newBegin.size() > end.size()) {
      return help(end, newBegin, bank, level);
    } else {
      return help(newBegin, end, bank, level);
    }
  }
};
